### ISSUE TITLE
Flutter should respect custom system-wide text selection toolbar buttons

### ISSUE THREAD

**Issue by justinmc** [flutter/flutter#107603]
On some platforms, it's possible to write an app that adds a custom button to the system-wide text selection toolbar.  For example, here the Android app AnkiDroid has added the "Anki Card" button to my text selection toolbar, and it appears in any app, such as the Google app here.

<img width="300" src="https://user-images.githubusercontent.com/389558/178827259-299e765b-e58c-4233-9bd4-367400058abe.png" />

Flutter should respect these added buttons, so that in any Flutter app, the buttons that the user has added to their device show up in the default text selection toolbar just like in any native app.

### Scope
This issue does not include allowing Flutter apps to add system-wide buttons.  It's about the reverse problem.

### Platforms

| Platform | Behavior | Related Work |
| --- | --- | --- |
| Android | Apps can add custom system-wide buttons ([relevant article](https://dev.to/bigaru/providing-custom-text-selection-actions-in-android-1akc)). Is there a way for Flutter to read what these buttons are and trigger them? | https://github.com/flutter/flutter/issues/139361 |
| iOS | ? | |
| Mac | ? | |
| Linux | ? | |
| Windows | ? | |

**Comment by gnprice** [comment]
It would be great to have this feature!

Copying what I wrote at https://github.com/flutter/flutter/issues/107578#issuecomment-1430782248, slightly edited:

As an Android user I find the "Translate" button in particular to be extremely useful, any time I'm reading text in a language other than English. (Even if I know the language moderately well, I'll use it regularly for words and phrases I don't understand.) When it's missing in a given app, if it's any app that's heavy on reading text, I'll notice and it's disappointing.

So as a developer I imagine many of my users value it too.

I believe that "Translate" button is added by the Google Translate app, and so falls in this category. (Other translation apps like DeepL add their own text selection buttons, too.)

**Comment by justinmc** [comment]
It looks like the "Translate" menu button does come from the Google Translate app.  I tried uninstalling the app and the button went away on my phone.

**Comment by gualse** [comment]
Just for fun, I was able to implement this feature in an empty project. I don't know how this implementation should be in Flutter itself but it doesn't seem too hard. You can see my code here: https://github.com/gualse/flutter_android_process_text/

**Comment by justinmc** [comment]
@gualse Wow that is awesome.  It correctly displays all the buttons on my phone and they all work.

Are you up for contributing this?  We'd need an engine PR that adds `textActionList` and `processTextAction` to the method channels, and then a framework PR to use them for the default context menu buttons.

<img width="300" src="https://github.com/flutter/flutter/assets/389558/8c377eef-197a-4572-825f-3d511a87f8e8" />
<img width="300" src="https://github.com/flutter/flutter/assets/389558/3596c5d9-e7c7-498a-977b-c0d4cb551799" />

CC @LouiseHsu.  I believe you've looked into this kind of thing on iOS.  Not sure about there but apparently it's doable on Android.

**Comment by LouiseHsu** [comment]
@justinmc let me look into this a little bit more - From what I've seen, I am not sure there is an ios equivalent of Android's PROCESS_TEXT functionality which would allow you to get access to custom actions added by other apps

**Comment by justinmc** [comment]
No problem!  Thanks for showing that it's possible.

**Comment by gualse** [comment]
@justinmc I have improved [my repo](https://github.com/gualse/flutter_android_process_text/) a bit. Now it closes the toolbar and supports editable actions that apps like [EncryptTextApp](https://github.com/JanmanX/EncryptTextApp/tree/master) provide.

Also, I would like to point out that:
- By default, the flutter toolbar expends to full width but the native toolbar always holds the buttons of user apps inside three dots.
- In modern versions of Android, the native toolbar is fully rounded and its text isn't bold.

**Comment by justinmc** [comment]
@gualse Thanks!  Can you file separate issues for those two bullet points?  I don't think I've seen them before.

**Comment by justinmc** [comment]
@bleroux Are you interested in implementing this?  It looks to be quite high impact for Android users and should only take a small PR in the engine and framework each...  If it's as easy as the code in @gualse's [repo](https://github.com/gualse/flutter_android_process_text/) makes it look.

**Comment by bleroux** [comment]
@justinmc
Sure! That's a very interesting work.
I will try to transpose @gualse implementation in the coming days and share my progress.

**Comment by gnprice** [comment]
Interesting. I wouldn't expect that this would need anything in the manifest that isn't also needed by native Android apps for the same functionality.

Looking at what appears to be the implementation of this menu in the Android framework:
https://android.googlesource.com/platform/frameworks/base/+/android-13.0.0_r67/core/java/android/widget/Editor.java#7755
https://android.googlesource.com/platform/frameworks/base/+/android-13.0.0_r67/core/java/android/widget/Editor.java#7832
it's invoking exactly the same API, making exactly the same call to `PackageManager.queryIntentActivities`. I believe that code runs in the app's JVM, so I'd expect the system isn't able to give it any special privileges compared to the app's own code when it comes to permissions and the need to declare things in advance in the app manifest. But I'm not an expert in this area and may be missing something.

What happens if you leave that element out of the manifest? Does the query stop finding the services it should find?

**Comment by bleroux** [comment]
@gnprice Many thank for the links to Android implementation!

> What happens if you leave that element out of the manifest? Does the query stop finding the services it should find?

For the moment, I tested this only on one device (Pixel 7), when I remove the `<queries>` element I get no result, when I add it, I get two results on this phone (Google Translate and probably Voice Over).

Reading the Andoid documentation for `<queries>`, several activities might be automatically visible: https://developer.android.com/training/package-visibility/automatic.
Maybe we can set up things to comply with one of these rules?

**Comment by gualse** [comment]
@gnprice `<queries>` literally need to find out 3rd-party apps that can handle `android.intent.action.PROCESS_TEXT` because since Android 11 (API 30) apps can't get the list of installed apps. And even a native app needs these lines, despite they don't exist by default in a new project and I have no idea why. The developers of Android may even have forgotten about `PROCESS_TEXT` because even Jetpack Compose still doesn't support this useful feature. Also, I know that `PROCESS_TEXT` was introduced in Android 6.0 (API 23), when there was no installed apps restriction. Maybe of that's why there is no system API that doesn't required add `<queries>`.

In my opinion, in Flutter `<queries>` lines should be in the manifest of a new project because it will allow other developers remove these lines if for some reason they don't want them. I don't want these lines to be added manually, because other developers may not know about this. (For example, I don't have any experience with iOS native development and I even never had my own iOS device, so I will be surprised if I have to add some lines in my app for a usual feature like this). I also don't see any reason not to do this, such as the Google Play restrictions. But I could be wrong, this is just my opinion.

**Comment by gnprice** [comment]
I think it's quite reasonable to add this to the manifest if we find that that's what we need to do in order to make this feature work. Probably that would mean putting it in the template for a new project (`packages/flutter_tools/templates/app_shared/android.tmpl/app/src/main/AndroidManifest.xml.tmpl`), and perhaps also having an auto-migration to try to add it to existing projects at `flutter build` time.

I'm still a bit puzzled and wonder if there's something we're missing, though.

> I also don't see any reason not to do this, such as the Google Play restrictions.

I don't see a downside to *having* that element in the manifest. But there is a reason it'd be nice to find a way not to *need* it in the manifest: the manifest is generated from a template and then lives in the app's source tree, and I believe we don't have a way to add something to it that will reliably work for all existing apps. So if we need something in the manifest, then the code will need to handle the possibility it's not there, and some existing apps (i.e. where `flutter create` was run before we added this feature) won't have it.

---

> when I remove the `<queries>` element I get no result, when I add it, I get two results on this phone (Google Translate and probably Voice Over).

Interesting, thanks for the experiment.

The thing that leaves me puzzled, then, is that from just looking around at apps on my phone, I'm not finding any that have an element like this in their `<queries>`, even when their text selection menus do have all the options I'd expect. For example the Firefox app — selecting some text, the menu includes options supplied by Google Translate and DeepL and other apps, but the `<queries>` only has `android.intent.action.SEND`. And I'm running Android 13, and the app's targetSdkVersion is 33, so that should be the latest behavior.

In other words this doesn't seem to always be true:

> And even a native app needs these lines

though perhaps in some circumstances it is.

---

> The developers of Android may even have forgotten about `PROCESS_TEXT` because even Jetpack Compose still doesn't support this useful feature.

This is interesting. @gualse do you mean that if you create a new app using Jetpack Compose, and it has some text in it and the user selects the text, they don't get these custom options?

**Comment by bleroux** [comment]
FYI, I filed https://github.com/flutter/engine/pull/44579 to gather some feedback on the overall approach.

@justinmc I have two questions in mind:
- do we stick to the Android naming ("process text")? (I have not investigate if similar API are available on the other platforms).
- should we consider that this feature might belong to flutter/packages?

**Comment by gualse** [comment]
@gnprice Yes, it is.

I created a new app, add the `<queries>` to the manifest. Then, I created `res/layout/edit_text.xml`:
```xml
<?xml version="1.0" encoding="utf-8"?>
<EditText xmlns:android="http://schemas.android.com/apk/res/android"
    android:id="@+id/editText"
    android:layout_width="0dp"
    android:layout_height="0dp"
    android:inputType="text" />
```
Then I changed the `onCreate` method:
```kotlin
    @OptIn(ExperimentalMaterial3Api::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            Column {
                val text = rememberSaveable { mutableStateOf("") }
                TextField(value = text.value, onValueChange = { text.value = it })

                // Caution: EditTexts without resource id don't have action buttons
                AndroidView(
                    modifier = Modifier.width(256.dp),
                    factory = { LayoutInflater.from(it).inflate(R.layout.edit_text, null) }
                )
            }
        }
    }
```
Then I ran it and only the native EditText had actions. (I don't even see a way to add any custom actions to `TextField`)

**Comment by camsim99** [comment]
> I don't see a downside to _having_ that element in the manifest. But there is a reason it'd be nice to find a way not to _need_ it in the manifest: the manifest is generated from a template and then lives in the app's source tree, and I believe we don't have a way to add something to it that will reliably work for all existing apps. So if we need something in the manifest, then the code will need to handle the possibility it's not there, and some existing apps (i.e. where `flutter create` was run before we added this feature) won't have it.

@justinmc I agree with this note. After digging into this a little bit, I'm not sure it's possible to add the element programmatically, and I don't see any downsides (but I would make 100% sure at least of this fact before adding it to the manifest template). Adding a note in the documentation about needing the manifest element should be ok for handling older projects and hopefully, we can add some error handling, too.

**Comment by justinmc** [comment]
@bleroux I say let's stick with the Android naming for now, and I think I'll also advocate for keeping this in Flutter proper instead of flutter/packages.  To me it seems like the vast majority of devs will want this out of the box.  Making this a package would probably result in lots of Flutter apps getting shipped without support for some user's favorite context menu button, just due to the develop not knowing they needed the package.

**Comment by justinmc** [comment]
Actually, what is the default behavior on native Android apps?  I just tried sticking a text input in a blank Android app and running it on my phone, and I don't see any of my third party app buttons.  Maybe I was wrong above and if Android is treating this as something that the developer needs to manually activate then we should too?

**Comment by Hixie** [comment]
If there's no downside to adding the `queries` block, I would probably add it anyway, as that will lead to a better use experience.

IIRC we have logic to automatically add the NETWORK permission during debug mode, we might be able to automatically add this too?

**Comment by gnprice** [comment]
@gualse Fascinating, thanks for the experiment. I reproduce your results. To be explicit:
* Without `<queries>` in the manifest, neither input has custom actions.
* With that `<queries>` element in the manifest, the native `EditText` has custom actions, but the Jetpack Compose `TextField` still lacks them.

Which also agrees with @justinmc's findings at https://github.com/flutter/flutter/issues/107603#issuecomment-1678205375 .

It's interesting that the Jetpack Compose `TextField` lacks this feature. That suggests that it's more of a ground-up rewrite of the Android SDK's component library than I'd realized. For all their end users' sake, I hope they do notice this missing feature and fill it in.

> The developers of Android may even have forgotten about `PROCESS_TEXT` because even Jetpack Compose still doesn't support this useful feature.

I think the Android team is very large, and so are the Android codebase and the overall Android feature set. Quite possibly nobody on the Jetpack Compose team knows about how this feature works, because there's so much else for them to think about. But somewhere off on another team within Android there are still probably others who are well aware of it.

---

There's still a mystery here, because as I mentioned [above](https://github.com/flutter/flutter/issues/107603#issuecomment-1670085219), some of the apps I use do get these custom actions in their text selection menus, and without having that query in a `<queries>` element. 

… Oh. Oh, I see. I think the point is that the `<queries>` element doesn't control what queries you can get results for; it only controls what packages you can see. You'll then see those packages when you make any arbitrary query that should match them.

And those other apps do have `<queries>` in their manifest, just not referring to `PROCESS_TEXT`. For example here's the cleanest one, from [the Firefox app](https://play.google.com/store/apps/details?id=org.mozilla.firefox) mentioned above:
```xml
    <queries>
        <intent>
            <action android:name="android.intent.action.SEND" />
            <data android:mimeType="text/plain" />
        </intent>
    </queries>
```

Lots of apps provide a `SEND` intent-filter; in fact, nearly all of the apps I've looked at that provide `PROCESS_TEXT` also provide `SEND`. That means they'll be visible to the Firefox app, for whatever queries it makes, including `PROCESS_TEXT`.

Confirming that understanding, if I put that version of `<queries>` into the test app, the `EditText` still gets nearly all the same options in its text selection menu, just like Firefox does; in particular it gets all the apps that provide `PROCESS_TEXT` and also provide `SEND`.

Moreover,  to eliminate any possibility that the `EditText` is responsible for doing something fancy: if I do a query for `PROCESS_TEXT` myself, just like the one in https://github.com/flutter/engine/pull/44579 :
```kt
                val intent = Intent().apply { action = Intent.ACTION_PROCESS_TEXT; type = "text/plain" }
                val resolved = packageManager.queryIntentActivities(intent, 0)
                Text(text = resolved.joinToString { it.activityInfo.packageName })
```
then I get the exact same list of packages that's reflected in the `EditText` text selection menu. That's true for each of the three `<queries>` branches of the experiment: no `<queries>`, a `<queries>` with `PROCESS_TEXT`, or a `<queries>` with `SEND`.

For us, the bottom line is that in order for the `PROCESS_TEXT` query to work, we'll indeed need something in `<queries>` in the manifest. And naturally it should be for the same query we actually want to make.

**Comment by bleroux** [comment]
I'm actively working on the Android side of the feature, see https://github.com/flutter/flutter/issues/139361.

Removing myself from this broader issue because there are no short term plan to implement this on the other platforms.

**Comment by gnprice** [comment]
@bleroux Great to hear! Looking forward to it.

FWIW I'm not sure this is a feature that any platforms other than Android have. Specifically:
* iOS: On an iOS device that has both Anki and Google Translate installed, if I select some text I don't get an item in the toolbar for either of those. (Both of the corresponding Android apps do create such a toolbar item.)
* Linux: There isn't even a system text-selection toolbar in the first place, at least not in my experience as a Linux user. (And I do use Gnome, which is the mainstream window manager.)
  * Just selecting text doesn't cause any kind of menu to appear.
  * Right-clicking often does show a menu, which is the closest analogy to the text selection toolbar of Android or iOS. But between Chrome, Firefox, Android Studio / IntelliJ, and the Gnome Terminal app, no two of those four right-click menus look alike. It seems like it's pretty much independently created by each app respectively.

So I'm pretty sure this feature doesn't exist on iOS or Linux. Not sure about macOS or Windows; I'm away from my Mac so can't immediately check there.

If the underlying feature is indeed Android-only, it'll probably make sense to close this issue once it's done for Android.

**Comment by bleroux** [comment]
@gnprice Interesting feedback, thanks.

Because I'm actively working on the Android implementation (and hope to complete it soon), I wanted to decouple the Android implementation from this issue because this one requires some researches to find out if this is applicable to other platforms.

Your comments related to iOS and Linux are very valuable, this is the kind of analysis we need to understand if this issue is Android only or not.

I'm too a Linux user, working from time to time on Mac or Windows. On macOS, there is a 'Services' sub-menu in the text selection menu, I don't know if applications can contribute to it (https://en.wikipedia.org/wiki/Services_menu).
